src/main/kotlin/io/kotlintest/KotlinTestEngine.kt

```diff
@@ -0,0 +1,220 @@
+      root.addChild(spec.specDescriptor)
+    }
+    return root
+  }
+}
+
+// a descriptor for an actual test case, allows config to be set which affects that test only
+class TestCaseDescriptor(private val id: UniqueId,
+                         private val displayName: String,
+                         private val source: TestSource,
+                         val spec: Spec,
+                         val test: () -> Unit,
+                         var config: TestCaseConfig) : TestDescriptor {
+
+  private var parent: Optional<TestDescriptor> = Optional.empty()
+
+  override fun getSource(): Optional<TestSource> = Optional.of(source)
+
+  override fun removeFromHierarchy() {
+    throw UnsupportedOperationException()
+  }
+
+  override fun setParent(parent: TestDescriptor?) {
+    this.parent = Optional.ofNullable(parent)
+  }
+
+  override fun getParent(): Optional<TestDescriptor> = parent
+
+  override fun getChildren(): MutableSet<out TestDescriptor> = mutableSetOf()
+
+  override fun getDisplayName(): String = displayName
+
+  override fun getType(): TestDescriptor.Type = TestDescriptor.Type.TEST
+
+  override fun getUniqueId(): UniqueId = id
+
+  override fun removeChild(descriptor: TestDescriptor?) {
+    throw UnsupportedOperationException()
+  }
+
+  override fun addChild(descriptor: TestDescriptor?) {
+    throw UnsupportedOperationException()
+  }
+
+  override fun findByUniqueId(uniqueId: UniqueId): Optional<out TestDescriptor> = if (id == uniqueId) Optional.of(this) else Optional.empty()
+
+  override fun getTags(): MutableSet<TestTag> = mutableSetOf()
+
+  fun config(
+      invocations: Int? = null,
+      enabled: Boolean? = null,
+      timeout: Duration? = null,
+      threads: Int? = null,
+      tags: Set<Tag>? = null,
+      interceptors: List<(TestCaseContext, () -> Unit) -> Unit>? = null) {
+    config =
+        TestCaseConfig(
+            enabled ?: config.enabled,
+            invocations ?: config.invocations,
+            timeout ?: config.timeout,
+            threads ?: config.threads,
+            tags ?: config.tags,
+            interceptors ?: config.interceptors)
+  }
+
+  internal val isActive: Boolean
+    get() = config.enabled && isActiveAccordingToTags
+
+  private val isActiveAccordingToTags: Boolean
+    get() {
+      val testCaseTags = config.tags.map { it.toString() }
+      val includedTags = readProperty("includeTags")
+      val excludedTags = readProperty("excludeTags")
+      val includedTagsEmpty = includedTags.isEmpty() || includedTags == listOf("")
+      return when {
+        excludedTags.intersect(testCaseTags).isNotEmpty() -> false
+        includedTagsEmpty -> true
+        includedTags.intersect(testCaseTags).isNotEmpty() -> true
+        else -> false
+      }
+    }
+
+  private fun readProperty(name: String): List<String> =
+      (System.getProperty(name) ?: "").split(',').map { it.trim() }
+}
+
+// the root descriptor
+class EngineTestDescriptor : TestDescriptor {
```

Comment:
This class seems to be unused? If it is used reflectively, this should be documented.